### ISSUE TITLE
clustermesh: missing identity from remote cluster leading to dropped traffic

### ISSUE THREAD

**Issue by rcanderson23** [cilium/cilium#32060]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

We have 6 clusters, all in different aws regions, configured with clustermesh. Our workloads are broken into two types `proxy` and `auth` with `auth` living in only one of the clusters and `proxy` living in all the clusters. We observed a `proxy` pod failing to start due to being unable to connect to its respective `auth` pod that lived in another cluster. Hubble showed traffic being dropped due to `Policy denied`. The `auth` IP was properly resolved but had the label `reserved:world`. Running `cilium identity list` on the affected cilium pod showed a missing identity for the target `auth`.

<details>
<summary>Unhealthy node:</summary>

```
106300     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-usw2
           k8s:role=proxy
--
154213     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-use1
           k8s:role=proxy
--
225920     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-euc1
           k8s:role=proxy
--
251430     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-euc1
           k8s:role=proxy
--
280542     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=tc-prod-cs-01-aps1
           k8s:role=proxy
--
350792     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-apse1
           k8s:role=proxy
--
398899     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-sae1
           k8s:role=proxy
```
</details>

<details>
<summary>Healthy node:</summary>

```
77498      k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-usw2
           k8s:role=auth
--
106300     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-usw2
           k8s:role=proxy
--
154213     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-use1
           k8s:role=proxy
--
225920     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-euc1
           k8s:role=proxy
--
251430     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-euc1
           k8s:role=proxy
--
280542     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-aps1
           k8s:role=proxy
--
350792     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-apse1
           k8s:role=proxy
--
398899     k8s:app=teleport-cloud
           k8s:io.cilium.k8s.namespace.labels.kubernetes.io/metadata.name=namespace1
           k8s:io.cilium.k8s.policy.cluster=cs-01-sae1
           k8s:role=proxy
```
</details>

The surprising portion is that the `proxy` from usw2 is present but not the `auth` in usw2. I expect that both of those are known from the etcd cluster present in usw2. Restarting the cilium pod resulted in the complete identity list being generated and the affected `proxy` being able to connect to `auth`.

I have collected a sysdump from the affected cilium pod but haven't posted as I would likely need to do some redaction. If there are specific files that are useful that would make it easier to provide the information required.

### Cilium Version

Client: 1.14.8 cf6e022e 2024-03-13T12:23:35-04:00 go version go1.21.8 linux/amd64
Daemon: 1.14.8 cf6e022e 2024-03-13T12:23:35-04:00 go version go1.21.8 linux/amd64

### Kernel Version

Linux ip-10-19-101-0.eu-central-1.compute.internal 6.1.82 #1 SMP PREEMPT_DYNAMIC Fri Apr  5 22:26:15 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### Kubernetes Version

Server Version: v1.28.8-eks-adc7111

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by rcanderson23** [comment]
We spotted an issue that feels similar to this in one of our environments. The identity was present but Service -> Pod IP mapping was missing. The only pods for the service existed in a different cluster.

```
# Bad state
cilium service list | grep 172.20.178.67
8     172.20.178.67:26258    ClusterIP                                           
9     172.20.178.67:26257    ClusterIP 

# Good state
cilium service list | grep 172.20.178.67     
953   172.20.178.67:26257    ClusterIP      1 => 10.18.127.195:26257 (active)    
954   172.20.178.67:26258    ClusterIP      1 => 10.18.127.195:26258 (active)
```

I have collected a sysdump of a node exhibiting this bad state as well and can provide redacted logs on request.

**Comment by rcanderson23** [comment]
We have hit this once again. We had about ~5 pods across 3 clusters run into this causing app issues. Restarting the cilium pods on the nodes resolves the issue.

**Comment by giorio94** [comment]
I'm mildly confident that the underlying cause should be fixed by https://github.com/cilium/cilium/pull/34091. Could you please give the latest patch release a try (the fix got backported to all affected releases)?

**Comment by rcanderson23** [comment]
> I'm mildly confident that the underlying cause should be fixed by #34091. Could you please give the latest patch release a try (the fix got backported to all affected releases)?

I believe you are right. We are running a custom version currently with https://github.com/cilium/cilium/pull/34148 patched which looks like the same fix you pushed and haven't encountered the issue again where we were previously able to consistently reproduce it. We are currently in the process of moving to latest v1.15 so we should be good there. Thanks!

**Comment by giorio94** [comment]
Perfect, thanks for confirming.
